src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -373,9 +387,46 @@ public Task UnloadAsync()
             return DisposeAsync();
         }
 
+        private async Task NotifySourceFileDirtyAsync(string projectRelativeSourceFileName)
+        {
+            DesignTimeInputsItem inputs = AppliedValue.Value;
+
+            // We check shared inputs first in case a file is marked as both, we don't want to compile twice
+            if (inputs.SharedInputs.Contains(projectRelativeSourceFileName))
+            {
+                foreach (string item in inputs.Inputs)
+                {
+                    await FireTempPEDirtyAsync(item, false);
+                }
+            }
+            else if (inputs.Inputs.Contains(projectRelativeSourceFileName))
+            {
+                await FireTempPEDirtyAsync(projectRelativeSourceFileName, true);
+            }
+        }
+
+        protected virtual async Task SubscribeToFileChangesAsync(ImmutableDictionary<string, uint>.Builder cookies, string projectRelativeSourceFileName)
+        {
+            IVsAsyncFileChangeEx fileChangeService = await _fileChangeService.GetValueAsync();
+
+            string fileName = UnconfiguredProject.MakeRooted(projectRelativeSourceFileName);
+
+            // We don't care about delete and add here, as they come through data flow, plus they are really bouncy - every file change is a Time, Del and Add event)
+            uint cookie = await fileChangeService.AdviseFileChangeAsync(fileName, _VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size, sink: this);
+            cookies.Add(projectRelativeSourceFileName, cookie);
+        }
+
+        protected virtual async Task UnsubscribeFromFileChangesAsync(uint cookie)
+        {
+            IVsAsyncFileChangeEx fileChangeService = await _fileChangeService.GetValueAsync();
+
+            await fileChangeService.UnadviseFileChangeAsync(cookie);
+        }
+
         protected override async Task DisposeCoreAsync(bool initialized)
         {
             DisposeTaskSchedulers();
+            await UnregisterFileWatchersAsync();
```

Comment:
There's a race; you could be in the middle of ApplyAsync when Dispose is called, this will unregister the file watchers but then ApplyAsync will add more.